diesel/src/sqlite/connection/row.rs

```diff
@@ -0,0 +1,155 @@
+    type Field = SqliteField<'a, 'b>;
+    type InnerPartialRow = Self;
+
+    fn field_count(&self) -> usize {
+        self.field_count
+    }
+
+    fn get<I>(&self, idx: I) -> Option<Self::Field>
+    where
+        Self: RowIndex<I>,
+    {
+        let idx = self.idx(idx)?;
+        Some(SqliteField {
+            row: SqliteRow {
+                inner: self.inner.clone(),
+                field_count: self.field_count,
+            },
+            col_idx: i32::try_from(idx).ok()?,
+            field_name: OnceCell::new(),
+        })
+    }
+
+    fn partial_row(&self, range: std::ops::Range<usize>) -> PartialRow<Self::InnerPartialRow> {
+        PartialRow::new(self, range)
+    }
+}
+
+impl<'a: 'b, 'b> RowIndex<usize> for SqliteRow<'a, 'b> {
+    #[inline]
+    fn idx(&self, idx: usize) -> Option<usize> {
+        if idx < self.field_count {
+            Some(idx)
+        } else {
+            None
+        }
+    }
+}
+
+impl<'a: 'b, 'b, 'd> RowIndex<&'d str> for SqliteRow<'a, 'b> {
+    fn idx(&self, field_name: &'d str) -> Option<usize> {
+        match &mut *self.inner.borrow_mut() {
+            PrivateSqliteRow::Direct(stmt) => stmt.index_for_column_name(field_name),
+            PrivateSqliteRow::Duplicated { column_names, .. } => column_names
+                .iter()
+                .position(|n| n.as_ref().map(|s| s as &str) == Some(field_name)),
+            PrivateSqliteRow::TemporaryEmpty => {
+                unreachable!()
+            }
+        }
+    }
+}
+
+#[allow(missing_debug_implementations)]
+pub struct SqliteField<'a, 'b> {
+    row: SqliteRow<'a, 'b>,
+    col_idx: i32,
+    field_name: OnceCell<Option<String>>,
+}
+
+impl<'a: 'b, 'b> Field<Sqlite> for SqliteField<'a, 'b> {
+    fn field_name(&self) -> Option<&str> {
+        self.field_name
+            .get_or_init(|| match &mut *self.row.inner.borrow_mut() {
+                PrivateSqliteRow::Direct(stmt) => {
+                    stmt.field_name(self.col_idx).map(|s| s.to_owned())
+                }
+                PrivateSqliteRow::Duplicated { column_names, .. } => column_names
+                    .get(self.col_idx as usize)
+                    .and_then(|n| n.clone()),
+                PrivateSqliteRow::TemporaryEmpty => unreachable!(),
+            })
+            .as_ref()
+            .map(|s| s as &str)
+    }
+
+    fn is_null(&self) -> bool {
+        self.value().is_none()
+    }
+
+    fn value<'d>(&'d self) -> Option<crate::backend::RawValue<'d, Sqlite>> {
+        match &*self.row.inner.borrow() {
+            PrivateSqliteRow::Direct(stmt) => stmt.value(self.col_idx),
+            PrivateSqliteRow::Duplicated { values, .. } => {
+                values.get(self.col_idx as usize).and_then(|v| {
+                    v.as_ref()
+                        .and_then(|v| unsafe { SqliteValue::new(v.value.as_ptr()) })
+                })
+            }
+            PrivateSqliteRow::TemporaryEmpty => unreachable!(),
+        }
+    }
```

Comment:
I'm not sure if that function is really sound, especially if we could get surprising behaviour somewhere if someone manages to cache a single value somewhere + we change the duplicated representation internally. A review would be welcome.

Reply:
As expected this was not sound at all. Replaced by a new better implementation that does not need to use unsafe.